Nix codegen

jonascarpay opened the pull request:
Currently outputs the following nix for Main.purs:
```nix
modules:
let
  useMySubstring = modules.Main.mySubstring 0 3 "nixos";
  myId = a: a;
  myNum = modules.Main.myId modules.Dep.myDep;
in
  { inherit myId myNum mySubstring useMySubstring;
    inherit (modules.Dep) myDep;
  }
```

- [x] Constructors and case-expressions
- [x] Strictness annotations
- [x] Support multiple files
  - [x] Codegen for multiple individual files
  - [x] Generate an overarching file that resolves imports
  - [x] Due to the way PS qualifies variables, even local top-level binders are pulled from imports. This shouldn't be a problem with a project file implementation like the one below, and it might be safest to leave it be, but it's still unnecessarily confusing.
  - [x] Make sure nested modules work and are properly escaped
- [x] Builtins
  - [x] handling of foreign imports is broken, as can be seen by mySubString being exported but not declared
- [x] Printing
  - [x] Make sure parenthesization works correctly for associativity (like `a (b c) d`)
  - [x] The entire module could use some refactoring, e.g. `parenthesize` -> `parenthesizeInner`

The overarching project file fixpoint might look something like
```nix
let
  modules = {
    Main = import ./Main.nix modules;
    Dep = import ./Dep.nix modules;
  };
in modules.Main
```

jonascarpay committed:
Initial scaffolding

jonascarpay committed:
Refactor Expr

jonascarpay committed:
Conversion types and scaffolding

jonascarpay committed:
Show file names when throwing errors

jonascarpay committed:
Add convenience constructors for Expr

jonascarpay committed:
Module converter

jonascarpay committed:
Module -> Attr set structure

jonascarpay committed:
Module conversion

jonascarpay committed:
Conversion up to constructors and case expressions

jonascarpay committed:
Printer types

jonascarpay committed:
Generalize BinOp

jonascarpay committed:
Attr sets take a list

jonascarpay committed:
Let-bindings take a list instead of a map

jonascarpay committed:
Pretty printing

jonascarpay committed:
Simple keyword checking

jonascarpay committed:
Don't quote strings, they already are

jonascarpay committed:
Fix let printing

jonascarpay committed:
Rename nix "imports" to "modules", since it can refer to the same module

jonascarpay committed:
Pretty printer tweaks

jonascarpay committed:
Don't escape selectors if they all satisfy isAlphaNum

jonascarpay committed:
Perform keyword and escape checks on let and attr bindings

cdepillabout commented on purenix/src/Lib.hs:
```diff
@@ -2,28 +2,35 @@ module Lib where
 defaultMain = do
+  Env.getArgs >>= print
   corefn <- readFile "../purescript-cabal-parser/output/Main/corefn.json"
+  -- TODO why is this lazy?
   let maybeValue = decode $ encodeUtf8 $ pack corefn
```
Maybe you're asking here why `pack` is `Data.Text.Lazy.pack` and not just `Data.Text.pack`?

`Data.Aeson.decode` takes a lazy `ByteString` as input, so I used `Data.Text.Lazy.pack`.

Although now when I pull up the aeson haddocks and scroll down the page a little more, I see there is a `decodeStrict` I could have used instead:

https://hackage.haskell.org/package/aeson-1.5.6.0/docs/Data-Aeson.html#v:decodeStrict

Also, I'm sure you can tell, but this file was basically just a quick hack to show how I imagined the `spago build` workflow to work.  Feel free to fix/change anything here.

cdepillabout committed:
Make our own Prelude in order to reduce imports.

cdepillabout committed:
Make sure that the CLI purescript we are using is the same as the PureScript library we use in purenix.

cdepillabout committed:
Add function for stripping annotations from a PureScript Module.

This is convenient for debugging PureScript modules.

It can be used like the following:

```diff
diff --git a/purenix/src/Lib.hs b/purenix/src/Lib.hs
index 62305ff..3e7f33a 100644
--- a/purenix/src/Lib.hs
+++ b/purenix/src/Lib.hs
@@ -13,7 +13,9 @@ import qualified Data.Text.Lazy.IO as TL
 import Language.PureScript.CoreFn.FromJSON (moduleFromJSON)
 import Nix.Convert (convert)
 import Nix.Print (renderExpr)
+import Nix.Util (stripAnnMod)
 import System.Directory (createDirectoryIfMissing)
+import Text.Pretty.Simple (pPrint)
 import qualified System.Environment as Env
 import qualified System.Exit as Sys

@@ -31,6 +33,7 @@ defaultMain = do
         Left err -> error err
         Right (_version, mdl) -> do
           putStrLn "successfully decoded purescript module"
+          pPrint (stripAnnMod mdl)
           nix <- either (Sys.die . T.unpack) (pure . renderExpr) $ convert mdl
           TL.putStrLn nix
```

cdepillabout committed:
Add explanation for where GenIdent is used.

cdepillabout committed:
Update list of Nix primops that we want to avoid in idents.

cdepillabout committed:
Remove strings from attribute names in the converter.

Imagine we have the following purescript code:

```purescript
xxx :: { "assert hello" :: String, bye :: String }
xxx = { "assert hello": "hello", bye: "bye" }
```

Before this commit, purenix would generate the following code:

```nix
modules:
let
  xxx =
    { ""assert hello"" = "hello";
      bye = "bye";
    };
  ...
```

With this commit, the following is generated:

```nix
modules:
let
  xxx =
    { "assert hello" = "hello";
      bye = "bye";
    };
  ...
```

cdepillabout committed:
Remove quotes from a selector in the converter.

PureScript's `prettyPrintObjectKey` will add quotes to a selector if
necessary.

However, the purenix pretty-printer will handle adding quotes if
necessary, so we explicitly remove quotes added by `prettyPrintObjectKey`.

Imagine we have the following PureScript function:

```purescript
yyy :: { "assert hello" :: String } -> String
yyy xxx = xxx."assert hello"
```

Before this change, this would generate code like the following:

```nix
modules:
let
  yyy = xxx: xxx.""assert hello"";
  ...
```

With this change, the following is generated:

```nix
modules:
let
  yyy = xxx: xxx."assert hello";
  ...
```

cdepillabout commented:
> Strictness annotations

I gave this some thought.

I think there is a lot of stuff in Nixpkgs that relies on laziness.  A couple examples:

- all the fixed-point stuff
- overlays
- callPackage-style argument passing (with `enable*` arguments)

I think it would be easiest if we treated our PureScript as lazy.  I think this might be possible.  I imagine we'll need to re-implement most of the PureScript Prelude anyway, so this doesn't seem too crazy.

My biggest worry is that there is some sort of compilation step (or optimization pass) before the corefn generation that relies on PureScript being strict.  I don't know of anything like this off the top of my head, but it is something to think about.

cdepillabout committed:
Make sure Nix keywords are quoted when used as attr key.

In Nix, it is legal to use keywords as attribute set keys as long as
they are quoted:

```
nix-repl> {"if" = 3;}."if"
```

This commit makes sure to quote keywords in selectors and attribute
keys.

PureScript like the following:

```purescript
xxx :: { assert :: String } -> String
xxx v = v.assert

yyy :: { assert :: String }
yyy = { assert: "hello" }
```

Will now produce the following:

```nix
modules:
let
  yyy =
  { "assert" = "hello";
  };
  xxx = v: v."assert";
  ...
```

cdepillabout commented:
I took a look at how the main PureScript compiler handles FFI files.

If you create a normal PureScript project that looks like the following:

```console
$ tree .
.
├── packages.dhall
├── spago.dhall
├── src
│   ├── Main.js
│   └── Main.purs
└── test
    └── Main.purs
```

Where `Main.purs` is:

```purescript
module Main where

lalala :: Int
lalala = 3

foreign import hello :: Int
```

and `Main.js` is:

```js
exports.hello = 1;
```

and you run `spago build`, you get an `outputs` directory that looks like this:

```console
$ tree output/
output/
├── cache-db.json
└── Main
    ├── externs.cbor
    ├── foreign.js
    └── index.js
```

where `index.js` is:

```js
// Generated by purs version 0.14.0
"use strict";
var $foreign = require("./foreign.js");
var lalala = 3;
module.exports = {
    lalala: lalala,
    hello: $foreign.hello
};
```

and `foreign.js` is the exact same as `Main.js` above.

------------------------------------------------------

Here's roughly what happens in the PureScript compiler:

1. The main `compile` command for `purs` (the PureScript compiler binary name) reads in all of the `.purs` files, and then tries to infer the names of the `.js` files:

    https://github.com/purescript/purescript/blob/729822bc115085efe5441522f69eaec7eb4ff4ff/app/Command/Compile.hs#L63-L67

    Here's the actual code that looks for the corresponding `.js` file:

    https://github.com/purescript/purescript/blob/729822bc115085efe5441522f69eaec7eb4ff4ff/src/Language/PureScript/Make.hs#L240-L257

2. There is some sort of "make action" definition that defines both a function for geneating code (`codegen`) as well as performing checks on the ffi files `ffiCodegen`.

    - `codegen`:

        https://github.com/purescript/purescript/blob/729822bc115085efe5441522f69eaec7eb4ff4ff/src/Language/PureScript/Make/Actions.hs#L221-L250

        The important line here looks to be the call to `J.moduleToJs`:

        https://github.com/purescript/purescript/blob/729822bc115085efe5441522f69eaec7eb4ff4ff/src/Language/PureScript/CodeGen/JS.hs#L45-L78

        This appears to add a node to the output JS AST when foreign definitions exist.

    - `ffiCodegen`:

        https://github.com/purescript/purescript/blob/729822bc115085efe5441522f69eaec7eb4ff4ff/src/Language/PureScript/Make/Actions.hs#L254-L268

        It look slike `ffiCodegen` performs some checks, and then basically just copies the `Main.js` file to `outputs/Main/foreign.js`.

3. These `codegen` and `ffiCodegen` functions are used when actually rebuilding modules:

    https://github.com/purescript/purescript/blob/729822bc115085efe5441522f69eaec7eb4ff4ff/src/Language/PureScript/Make.hs#L67-L121

    First `ffiCodegen` is called, and then `codegen`.

jonascarpay committed:
Merge pull request #6 from cdepillabout/cdepillabout/codegen

cdepillabout committed:
Read in the FFI file when it exists.

cdepillabout committed:
Implement basic FFI support.

cdepillabout committed:
Get rid of unused import.

cdepillabout committed:
Make sure exports and ffi exports are correctly setup.

cdepillabout committed:
Remove unused imports.

cdepillabout commented:
> Due to the way PS qualifies variables, even local top-level binders are pulled from imports. This shouldn't be a problem with a project file implementation like the one below, and it might be safest to leave it be, but it's still unnecessarily confusing.

I was thinking that one potential problem with this is that it won't correctly reference top-level binders that are _not_ exported.

Top-level binders that are not exported will have to be pulled from the same module (without the `modules.Main.` prefix).

For instance, imagine the following PureScript module:

```purescript
module Main (foobar) where

foobar :: String
foobar = nonExportedIdent

nonExportedIdent :: String
nonExportedIdent = "hello"
```

Currently, this will get converted to:

```nix
modules: 
let
  nonExportedIdent = "hello";
  foobar = modules.Main.nonExportedIdent;
in
  { inherit foobar;
  }
```

But really it needs to be:

```nix
modules: 
let
  nonExportedIdent = "hello";
  foobar = nonExportedIdent;
in
  { inherit foobar;
  }
```

------------------------------------------------------------

Although personally, I'd also be fine with something like the following:

```nix
modules: 
let
  nonExportedIdent = "hello";
  foobar = modules.Main._allNonExported.nonExportedIdent;
  # this is a special identifier that packages up all non-exported bindings and exposes them for debugging purposes (or users who like to live dangerously)
  __allNonExported = { inherit nonExportedIdent; };
in
  { inherit foobar __allNonExported;
  }
```

jonascarpay committed:
Merge pull request #7 from cdepillabout/cdepillabout/ffi

jonascarpay commented on purenix/src/Nix/Convert.hs:
```diff
@@ -0,0 +1,174 @@
+    purenixIdents = ["modules", "__ffi"]
+    -- primops (builtins) in Nix that can be accessed without importing anything.
+    -- These were discovered by running `nix repl` and hitting TAB.
+    nixPrimops =
```
At this point it probably makes sense to turn this into a `Set` for cheap lookups

cdepillabout replied:
Yeah, that completely makes sense.

jonascarpay committed:
Constructors

cdepillabout committed:
Switch to importing ffi file instead of embedding it raw.

cdepillabout committed:
Make sure ffi file is output in generated/ dir.

jonascarpay committed:
Merge pull request #8 from cdepillabout/cdepillabout/ffi-again

jonascarpay committed:
WIP case statements

cdepillabout commented:
I was curious as to how the JS backend handles sum types and case statements.

Imagine you have the following PureScript file:

```purescript
module Main where

data FooBarBaz a
  = Foo
  | Bar Int
  | Baz a String Number
```

This gets compiled into the following JavaScript file:

```javascript
var Foo = (function () {
    function Foo() {

    };
    Foo.value = new Foo();
    return Foo;
})();
var Bar = (function () {
    function Bar(value0) {
        this.value0 = value0;
    };
    Bar.create = function (value0) {
        return new Bar(value0);
    };
    return Bar;
})();
var Baz = (function () {
    function Baz(value0, value1, value2) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
    };
    Baz.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return new Baz(value0, value1, value2);
            };
        };
    };
    return Baz;
})();
module.exports = {
    Foo: Foo,
    Bar: Bar,
    Baz: Baz
};
```

I played around with this in the nodejs repl a little:

```javascript
$ node
> Main = require('./output/Main/index.js');
{
  Foo: [Function: Foo] { value: Foo {} },
  Bar: [Function: Bar] { create: [Function (anonymous)] },
  Baz: [Function: Baz] { create: [Function (anonymous)] }
}
```

It looks like you can construct a `Foo` with `Foo.value`:

```javascript
> myFoo = Main.Foo.value
> myFoo
Foo {}
```

You can construct a `Bar` and `Baz` by using their curried constructors:

```javascript
> myBar = Main.Bar.create(100)
> myBar
Bar { value0: 100 }
> myBar.value0
100
> myBaz = Main.Baz.create([])("hello")(3.3)
> myBaz
Baz { value0: [], value1: 'hello', value2: 3.3 }
> myBaz.value1
'hello'
```

It looks like constructors effectively become something like a class or a named/tagged record?  I'm not really sure how JavaScript works or what these things are.  But they do appear to be like records, where you can access the fields with the normal selector syntax.

--------------------------------------------------------------

I then added a `case` statement to my PureScript file:

```purescript
myCaseTest :: FooBarBaz Int -> Int
myCaseTest fbb =
  case fbb of
    Foo -> 999
    Bar i -> i
    Baz i "hello" _ -> i
    Baz _ _ _ -> 0
```

This gets generated as:

```javascript
var myCaseTest = function (fbb) {
    if (fbb instanceof Foo) {
        return 999;
    };
    if (fbb instanceof Bar) {
        return fbb.value0;
    };
    if (fbb instanceof Baz && fbb.value1 === "hello") {
        return fbb.value0;
    };
    if (fbb instanceof Baz) {
        return 0;
    };
    throw new Error("Failed pattern match at Main (line 19, column 3 - line 23, column 19): " + [ fbb.constructor.name ]);
};
```

This seems to be pretty straight-forward.  I'm guessing `instanceof` is used to check the name/tag of each record.

I was able to use it like this:

```javascript
> Main.myCaseTest(myFoo)
999
> Main.myCaseTest(myBar)
100
> Main.myCaseTest(myBaz)
[]
> myOtherBaz = Main.Baz.create([])("bye")(3.3)
> Main.myCaseTest(myOtherBaz)
0
```

jonascarpay commented:
My own test with more complicated patterns:
```purescript
module Main where

import Prelude

data Maybe a = Nothing | Just a

data FBB a
  = Foo
  | Bar a
  | Baz a a

myCase :: FBB (Maybe Int) -> Int
myCase fbb = case fbb of
  Foo -> 0
  Bar Nothing -> 1
  Bar (Just n) -> n + 1
  Baz (Just a) (Just b) | a + b < 9 -> 9
  Baz _ Nothing -> 10
  Baz Nothing _ -> 11
  Baz (Just _) (Just _) | otherwise -> 12

myMultiCase :: Maybe Int -> Maybe Int -> Int
myMultiCase a b = case a, b of
  Nothing, Nothing -> 0
  Just n, Nothing -> n
  Just n, Just m | n < 9 -> n + m
  _, _ -> 2
```

Compiles to

```javascript
// Generated by purs version 0.14.4
"use strict";
var Data_Boolean = require("../Data.Boolean/index.js");
var Nothing = (function () {
    function Nothing() {

    };
    Nothing.value = new Nothing();
    return Nothing;
})();
var Just = (function () {
    function Just(value0) {
        this.value0 = value0;
    };
    Just.create = function (value0) {
        return new Just(value0);
    };
    return Just;
})();
var Foo = (function () {
    function Foo() {

    };
    Foo.value = new Foo();
    return Foo;
})();
var Bar = (function () {
    function Bar(value0) {
        this.value0 = value0;
    };
    Bar.create = function (value0) {
        return new Bar(value0);
    };
    return Bar;
})();
var Baz = (function () {
    function Baz(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Baz.create = function (value0) {
        return function (value1) {
            return new Baz(value0, value1);
        };
    };
    return Baz;
})();
var myMultiCase = function (a) {
    return function (b) {
        if (a instanceof Nothing && b instanceof Nothing) {
            return 0;
        };
        if (a instanceof Just && b instanceof Nothing) {
            return a.value0;
        };
        if (a instanceof Just && (b instanceof Just && a.value0 < 9)) {
            return a.value0 + b.value0 | 0;
        };
        return 2;
    };
};
var myCase = function (fbb) {
    if (fbb instanceof Foo) {
        return 0;
    };
    if (fbb instanceof Bar && fbb.value0 instanceof Nothing) {
        return 1;
    };
    if (fbb instanceof Bar && fbb.value0 instanceof Just) {
        return fbb.value0.value0 + 1 | 0;
    };
    if (fbb instanceof Baz && (fbb.value0 instanceof Just && (fbb.value1 instanceof Just && (fbb.value0.value0 + fbb.value1.value0 | 0) < 9))) {
        return 9;
    };
    if (fbb instanceof Baz && fbb.value1 instanceof Nothing) {
        return 10;
    };
    if (fbb instanceof Baz && fbb.value0 instanceof Nothing) {
        return 11;
    };
    if (fbb instanceof Baz && (fbb.value0 instanceof Just && (fbb.value1 instanceof Just && Data_Boolean.otherwise))) {
        return 12;
    };
    throw new Error("Failed pattern match at Main (line 13, column 14 - line 20, column 42): " + [ fbb.constructor.name ]);
};
module.exports = {
    Nothing: Nothing,
    Just: Just,
    Foo: Foo,
    Bar: Bar,
    Baz: Baz,
    myCase: myCase,
    myMultiCase: myMultiCase
};
```

jonascarpay committed:
Pattern matching

jonascarpay committed:
Fix parenthesization

jonascarpay committed:
Clean up string handling

jonascarpay committed:
Fix attr parenthesization

jonascarpay committed:
Put single field attr sets on a single line

jonascarpay committed:
Fix indexing of sum constructors

jonascarpay committed:
Drop `?` binary expr

jonascarpay committed:
Fix binary operator parenthesization

jonascarpay committed:
Merge pattern binds and scrutinee declarations

jonascarpay committed:
Make let bindings NonEmpty

jonascarpay committed:
Improve formatting for conditionals

jonascarpay committed:
Fix pattern matching error for always-pass patterns

jonascarpay committed:
Don't print FFI if it's not used

jonascarpay committed:
Define Path precedence

jonascarpay committed:
Remove outdated docs

jonascarpay committed:
Add haddocks for `alternative`

jonascarpay committed:
Print inherit-only attr sets on one line

jonascarpay committed:
Don't defer scrutinees

jonascarpay committed:
Typo in comment

Co-authored-by: Dennis Gosnell <cdep.illabout@gmail.com>

jonascarpay committed:
Reword/add haddocks

jonascarpay committed:
Update purenix/src/Nix/Expr.hs

Co-authored-by: Dennis Gosnell <cdep.illabout@gmail.com>

jonascarpay committed:
Add support for associative parenthesization

jonascarpay committed:
Don't export `__ffi`

jonascarpay committed:
Fix warnings

jonascarpay committed:
Pattern matching on objects

jonascarpay committed:
rename go -> litBinder

jonascarpay committed:
Print source of pattern match failure

jonascarpay committed:
Update purenix/src/Nix/Convert.hs

Co-authored-by: Dennis Gosnell <cdep.illabout@gmail.com>

jonascarpay committed:
Merge pull request #12 from cdepillabout/constructors

jonascarpay committed:
Inline imports and properly don't qualify local variables

jonascarpay committed:
Generate nix files in place

jonascarpay committed:
Allow more identifiers to be shadowed

jonascarpay committed:
Identifier overhaul

jonascarpay committed:
Remove Nix.Util

jonascarpay committed:
Always point foreign imports to ./foreign.nix

jonascarpay committed:
Filter out `Prim` and self imports

jonascarpay committed:
Use same rules for identifier -> key as identifier -> var

jonascarpay committed:
Copy FFI files

jonascarpay committed:
Always quote module name keys

jonascarpay committed:
Drop leading dollars rather than escaping them

jonascarpay committed:
stringToKey -> stringKey

Co-authored-by: Dennis Gosnell <cdep.illabout@gmail.com>

cdepillabout committed:
Handle newtypes.

jonascarpay committed:
Merge pull request #17 from cdepillabout/handle-newtypes

Handle newtypes
